### ISSUE TITLE
How do we manage the order of updates when using a filtered unique index?

### ISSUE THREAD

**Issue by sebdesalvador** [dotnet/efcore#35159]
## The question
I have a filtered unique index defined as:

```C#
builder
    .HasIndex(e => new { e.UserId, e.IsDefault })
    .HasFilter($"[{nameof(MyEntity.IsDefault)}] = 1")
    .IsUnique()
    .HasDatabaseName($"UX_{nameof(MyEntity)}_{nameof(MyEntity.UserId)}_{nameof(MyEntity.IsDefault)}");
```

It translates into:

```SQL
CREATE UNIQUE NONCLUSTERED INDEX [UX_MyEntity_UserId_IsDefault] ON [dbo].[MyEntity]
(
	[UserId] ASC,
	[IsDefault] ASC
)
WHERE ([IsDefault]=(1))
```

Because I can only have 1 record with the flag `IsDefault` set to `true`, the order of updates is important, I first need to set non default records to `false`, and then set the default one. Since the EF Core doesn't really know the condition/filter, it doesn't order the updates correclty and I often get an error with the message `Cannot insert duplicate key row in object 'dbo.MyEntity' with unique index 'UX_MyEntity_UserId_IsSelected'.`.

I have started experimenting with interceptors but it's getting dirty, surely someone had this issue before me?

Thanks!

### Provider and version information

EF Core version: 8.0.11
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 8.0
Operating system: Windows
IDE: Rider

**Comment by ajcvickers** [comment]
@AndriySvyryd This is a consequence of us not being able to parse the filter on the index. We did https://github.com/dotnet/efcore/pull/28923 a couple of years ago to avoid cycle breaking, but I don't see anything else tracking further changes to this. Do we have anything on the backlog that I missed?

**Comment by AndriySvyryd** [comment]
I don't think this is tracked anywhere else. We'd need to add API to provide a callback that determines what dependency should be broken in a cycle.

**Comment by roji** [comment]
Another possible answer here (for today) is to manually break the batch into two SaveChanges. This does cause an additional roundtrip, but anything else here seems like it could be too complicated/convoluted...

**Comment by rvdginste** [comment]
> Another possible answer here (for today) is to manually break the batch into two SaveChanges. This does cause an additional roundtrip, but anything else here seems like it could be too complicated/convoluted...

This reminds me of a similar problem you can get with a unique constraint that does not have a filter. When “reusing” the same unique key, you want to enforce the ‘delete’ happening before the ‘insert’. On nhibernate I used the technique mentioned by @roji . But should I interpret comment from @ajcvickers that efc enforces correct ordering automatically in that case?

**Comment by roji** [comment]
@rvdginste yes, it does.
